grails-app/services/org/pih/warehouse/product/ProductMergeService.groovy

```diff
@@ -0,0 +1,375 @@
+            }
+
+            logProductMergeData(primary, obsolete, obsoletedSupplier)
+
+            // Swap assemblyProduct to primary
+            obsoletedSupplier.product = primary
+            // Note: needs flush because of "User.locationRoles not processed by flush"
+            obsoletedSupplier.save(flush: true)
+        }
+
+        // 5. Synonym <-- if the same synonym does not exist on primary product, then can be swapped
+        List<Synonym> obsoletedSynonyms = Synonym.findAllByProduct(obsolete)
+        obsoletedSynonyms?.each { Synonym obsoletedSynonym ->
+            // Find if primary product has already the same synonym
+            Synonym primarySynonym = Synonym.findByProductAndName(primary, obsoletedSynonym.name)
+
+            if (primarySynonym) {
+                // if product synonym already exists, then do nothing with it
+                return
+            }
+
+            logProductMergeData(primary, obsolete, obsoletedSynonym)
+
+            // Swap assemblyProduct to primary
+            obsoletedSynonym.product = primary
+            // Note: needs flush because of "User.locationRoles not processed by flush"
+            obsoletedSynonym.save(flush: true)
+        }
+
+        // 6. Documents
+        obsolete.documents?.each { Document obsoleteDocument ->
+            logProductMergeData(primary, obsolete, obsoleteDocument)
+            obsolete.removeFromDocuments(obsoleteDocument)
+            primary.addToDocuments(obsoleteDocument)
+        }
+
+        /**
+         * INVENTORY ITEMS PLUS "ITEM" RELATIONS
+         * */
+
+        // 1. Copy Inventory items that are not already existing on the primary product (and InventorySnapshot change)
+        Set<InventoryItem> obsoleteInventoryItems = obsolete.inventoryItems
+        obsoleteInventoryItems?.each { InventoryItem obsoleteInventoryItem ->
+            // Check if this lot already exists for the primary product
+            InventoryItem primaryInventoryItem = InventoryItem.findByProductAndLotNumber(primary, obsoleteInventoryItem.lotNumber)
+
+            List<InventorySnapshot> inventorySnapshots = InventorySnapshot.findAllByInventoryItem(obsoleteInventoryItem)
+            List<ProductAvailability> productAvailabilities = ProductAvailability.findAllByInventoryItem(obsoleteInventoryItem)
+            if (primaryInventoryItem) {
+                // if product inventory item already exists, then check if obsolete InventoryItem has InventorySnapshot
+                if (inventorySnapshots) {
+                    inventorySnapshotService.updateInventorySnapshots(primaryInventoryItem, obsoletInventoryItem, primary)
+                }
+                // if product inventory item already exists, then check if obsolete InventoryItem has ProductAvailability
+                if (productAvailabilities) {
+                    productAvailabilityService.updateProductAvailabilityOnProductMerge(primaryInventoryItem, obsoletInventoryItem, primary)
```

Comment:
@jmiranda I moved the refresh product availability to here with the same approach as with the InventorySnapshot update.